hrm-domain/hrm-service/migrations/20230824113300-contact-function.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright (C) 2021-2023 Technology Matters
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+'use strict';
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize
+      .query(`CREATE OR REPLACE FUNCTION "contactRelations"(character varying(255), bigint)
+RETURNS TABLE (
+  "csamReports" jsonb,
+  "referrals" jsonb,
+  "conversationMedia" jsonb
+) AS $$
+  SELECT reports."csamReports", joinedReferrals."referrals", media."conversationMedia" FROM (
+    SELECT COALESCE(jsonb_agg(to_jsonb(r)), '[]') AS "csamReports"
+    FROM "CSAMReports" r
+    WHERE r."contactId" = $2 AND r."accountSid" = $1 AND r."acknowledged" = TRUE
+  ) reports
+  LEFT JOIN LATERAL (
+    SELECT COALESCE(jsonb_agg(to_jsonb(referral)), '[]') AS "referrals"
+    FROM "Referrals" referral
+    WHERE referral."contactId" = $2 AND referral."accountSid" = $1
+  ) joinedReferrals ON true
+  LEFT JOIN LATERAL (
+    SELECT COALESCE(jsonb_agg(to_jsonb(cm)), '[]') AS "conversationMedia"
+    FROM "ConversationMedias" cm
+    WHERE cm."contactId" = $2 AND cm."accountSid" = $1
+  ) media ON true
+$$
+LANGUAGE SQL
+STABLE;
```

Comment:
Marking the functions 'STABLE' should mean that the Postgres query planner should be able to optimise it's query plan for the whole query, taking the logic inside the function and the logic of the calling statement and treating it as one big query, meaning it shouldn't be any less performant than not using a function